Brokerages/Bitfinex/BitfinexWebsocketsBrokerage.Messaging.cs

```diff
@@ -0,0 +1,215 @@
+                        return;
+                    }
+                    else if (term == "tu" || term == "te")
+                    {
+                        //trade execution/update
+                        var data = raw[2].ToObject(typeof(string[]));
+                        PopulateTrade(data);
+                    }
+                    else if (term == "ws")
+                    {
+                        //wallet
+                        var data = raw[2].ToObject(typeof(string[][]));
+                        PopulateWallet(data);
+                    }
+                    else if (_channelId.ContainsKey(id) && _channelId[id] == "ticker")
+                    {
+                        //ticker
+                        PopulateTicker(e.Data);
+                        return;
+                    }
+                }
+                else if (raw.channel == "ticker" && raw.@event == "subscribed")
+                {
+                    if (!this._channelId.ContainsKey((int)raw.chanId))
+                    {
+                        this._channelId[(int)raw.chanId] = "ticker";
+                    }
+                }
+                else if (raw.chanId == 0)
+                {
+                    if (raw.status == "FAIL")
+                    {
+                        throw new Exception("Failed to authenticate with ws gateway");
+                    }
+                    Log.Trace("BitfinexWebsocketsBrokerage.OnMessage(): Successful wss auth");
+                }
+                else if (raw.@event == "info" && raw.code == "20051")
+                {
+                    //hard reset
+                    this.Reconnect();
+                }
+                else if (raw.@event == "info" && raw.code == "20061")
+                {
+                    //soft reset
+                    UnAuthenticate();
+                    Unsubscribe(null, null);
+                    Subscribe(null, null);
+                    Authenticate();
+                }
+
+                Log.Trace("BitfinexWebsocketsBrokerage.OnMessage(): " + e.Data);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, string.Format("Parsing wss message failed. Data: {0}", e.Data));
+            }
+        }
+        
+        private void PopulateTicker(string response)
+        {
+            var data = JsonConvert.DeserializeObject<string[]>(response, settings);
+            var msg = new TickerMessage(data);
+            lock (Ticks)
+            {
+                Ticks.Add(new Tick
+                {
+                    AskPrice = msg.ASK / ScaleFactor,
+                    BidPrice = msg.BID / ScaleFactor,
+                    AskSize = (long)Math.Round(msg.ASK_SIZE * ScaleFactor, 0),
+                    BidSize = (long)Math.Round(msg.BID_SIZE * ScaleFactor, 0),
+                    Time = DateTime.UtcNow,
+                    Value = msg.LAST_PRICE / ScaleFactor,
```

Comment:
Not sure the action to take here. I guess it would be possible to emit two ticks for each type, but this might lead to unexpected algorithm behavior. The core changes required to do the refactor would be very sensitive.